migrations/2021-04-28-154439_denormalize_action_receipt_actions/up.sql

```diff
@@ -0,0 +1,27 @@
+ALTER TABLE action_receipt_actions
+    ADD COLUMN predecessor_account_id text NOT NULL DEFAULT '',
+    ADD COLUMN receiver_account_id text NOT NULL DEFAULT '',
+    ADD COLUMN included_in_block_timestamp numeric(20, 0) NOT NULL DEFAULT 0;
+
+
+UPDATE action_receipt_actions
+    SET predecessor_account_id = receipts.predecessor_account_id,
+        receiver_account_id = receipts.receiver_account_id,
+        included_in_block_timestamp = receipts.included_in_block_timestamp
+    FROM receipts
+    WHERE action_receipt_actions.receipt_id = receipts.receipt_id;
+
+
+ALTER TABLE action_receipt_actions
+    ALTER COLUMN predecessor_account_id DROP DEFAULT,
+    ALTER COLUMN receiver_account_id DROP DEFAULT,
+    ALTER COLUMN included_in_block_timestamp DROP DEFAULT;
+
+CREATE INDEX action_receipt_actions_args_function_call_idx ON action_receipt_actions((args->>'method_name'))
+    WHERE action_receipt_actions.action_kind = 'FUNCTION_CALL';
+
+CREATE INDEX action_receipt_actions_args_amount_idx ON action_receipt_actions((args->'args_json'->>'amount'))
+    WHERE action_receipt_actions.action_kind = 'FUNCTION_CALL' AND (action_receipt_actions.args->>'args_json') IS NOT NULL;
```

Comment:
Reminder: we need to migrate the existing args to have args_json

Reply:
I don't like the idea to make such migration on the SQL side. Do you mind if I create additional command for that? I realise it will hung with us there forever, but I prefer to make such things via Rust rather that SQL.

Reply:
I’m fine with either SQL or Rust script

Reply:
@khorolets hmm, not sure why it's an issue  Postgres has no problem parsing base64, e.g.:

```sql
select * from (
    select date,
         -- TODO: How to filter only successful?
        count(distinct args->>'receiver_id') as receivers,
        sum((args->>'amount')::numeric) / 1e24 as volume
    from (
        select receipt_id,
            predecessor_account_id,
            receiver_account_id,
            args->>'method_name' as method_name,
            convert_from(decode(args->>'args_base64', 'base64'), 'UTF8')::json as args,
            date_trunc('week', to_timestamp(included_in_block_timestamp / 1000000000)) as date
        from action_receipt_actions join receipts using(receipt_id)
        where predecessor_account_id = :contract_name and receiver_account_id = 'wrap.near'
    ) as with_fields
    where method_name = 'ft_transfer'
    group by date
) as call_stats
order by date asc limit 100
```

so to migrate args via SQL you just need something like

```sql
UPDATE action_receipt_actions
SET args_base64 =  convert_from(decode(args->>'args_base64', 'base64'), 'UTF8')::json
```

which you can add to existing `UPDATE ... FROM` statement you have for other columns – that should be the fastest way to pull it off (loading to Rust and pushing back likely gonna take forever).

Reply:
@vgrichina thanks, I've tried but the migration which tried to decode failed on the records where `args_base64` was actual borsh args and it stopped migration. Do you have any suggestions how we can ignore those failing records and don't interrupt the process?

Reply:
@khorolets yeah, looks like it's a bit tricky but you can handle errors in Postgres:

```plpgsql
CREATE OR REPLACE FUNCTION decode_or_null(bytea) RETURNS json
   LANGUAGE plpgsql AS
$$BEGIN
   RETURN convert_from($1, 'UTF8')::json;
EXCEPTION
   WHEN others THEN
      RAISE WARNING '%', SQLERRM;
      RETURN NULL;
   
END;$$;

select decode_or_null(decode(args->>'args_base64', 'base64')) as args from action_receipt_actions limit 1000;
```

basically you can have your own wrapper function which returns NULL on every exception